### ISSUE TITLE
Kotlin Native Support

### ISSUE THREAD

**Issue by aajtodd** [awslabs/aws-sdk-kotlin#229]
Support for Kotlin Native targets

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue, please leave a comment

## Describe the Feature

Add support for using generated AWS service clients on Kotlin Native targets.
Requires:
* CRT support from aws-crt-kotlin
* Runtime support from smithy-kotlin (notably JSON/XML serde support)
* KMP project generation: https://github.com/awslabs/smithy-kotlin/issues/437

**Comment by napperley** [comment]
There is a typo in the description, **Add support for using generated AWS service clients on Kotlin JS targets**. Should be **Kotlin Native** instead of **Kotlin JS**.

**Comment by napperley** [comment]
Which targets will the Kotlin Native support cover? Does it include **linuxX64**, **linuxArm32Hfp**, and **linuxArm64**?

**Comment by aajtodd** [comment]
@napperley It's still a bit early to know for sure which targets we would initially support. I don't think any of the ones you mentioned are particularly worrisome though.

**Comment by aajtodd** [comment]
To anyone following this issue, we ask that you please take some time to respond to the [discussion](https://github.com/awslabs/aws-sdk-kotlin/discussions/464) on KMP use cases. Thank-you!

**Comment by AlbertoSH** [comment]
Hey, I'd love to see native support for the SDK. That would be a huge game changer when talking about microservices development (my main pain point is that yes, I can go for a native backend framework with Kotlin but I still need the clients for several services to be native)
As you were mentioning, [previous memory manager model has changed](https://blog.jetbrains.com/kotlin/2020/07/kotlin-native-memory-management-roadmap/). A preview of the new model[ was released last year](https://blog.jetbrains.com/kotlin/2021/08/try-the-new-kotlin-native-memory-manager-development-preview) and now [it's the default one for Kotlin 1.7.20](https://kotlinlang.org/docs/native-memory-manager.html) (as of today, latest version)

Everything to say, [there seems to be some performance issues with the new model](https://youtrack.jetbrains.com/issue/KT-48526), though. In any case, I prefer to have slow binaries that will improve with newer versions than having no binaries at all  ¯\\\_(ツ)\_/¯

I'll try to find some time to see if I can play around with the repo and generate some binary targets, but I'm afraid I can't promise anything

**Comment by rafalwrzeszcz** [comment]
Is there any status update on this? Kotlin Native would be a real game-changer with native Lambda runtime.

**Comment by aajtodd** [comment]
The team is focused on stabilizing the SDK right now for JVM. The core runtime is setup as a multiplatform library already though and ready to add support for additional platforms but we have no timeline to share at the moment.

**Comment by trueangle** [comment]
@rafalwrzeszcz couple months ago posted my own implementation of Kotlin/Native lambda runtime. Worth checking out https://github.com/trueangle/kotlin-native-aws-lambda-runtime

**Comment by rafalwrzeszcz** [comment]
> @rafalwrzeszcz couple months ago posted my own implementation of Kotlin/Native lambda runtime. Worth checking out https://github.com/trueangle/kotlin-native-aws-lambda-runtime

This issue is about SDK, not Lambda runtime

**Comment by jg-develop** [comment]
Is there any update on the ETA for native iOS support? It would be fantastic to be able to share common code for Android+iOS. A rough ETA would be helpful to know if it's worth waiting or going to the trouble of a separate iOS implementation while we wait. Thanks very much for your work - a really useful project...

**Comment by lauzadis** [comment]
Hi @jg-develop, this issue encompasses iOS support. This project is our top priority at the moment, it should be done later this year.

**Comment by carloshs94** [comment]
Hi @lauzadis, any more precise estimate on the ETA at the moment or is it still not ready to share updates?

**Comment by steamstreetjon** [comment]
Would really love to deploy some kotlin native lambdas. Are we close?

**Comment by lauzadis** [comment]
@steamstreetjon, @carloshs94, and everyone else following along -- thanks for your patience, and apologies for our silence here. The implementation has been complete and ready to release since early November. We already publish Kotlin/Native artifacts for [smithy-kotlin](https://github.com/awslabs/smithy-kotlin) and [aws-crt-kotlin](https://github.com/aws/aws-crt-kotlin), which support the aws-sdk-kotlin product.

Unfortunately, releasing aws-sdk-kotlin has taken longer than expected due to some publishing challenges related to the size of our native artifacts. We've been working closely with our publishing dependency to resolve these, and have identified a number of optimizations over the past month. Optimistically, I'm expecting this will be resolved within the next few weeks, but it could take a bit longer.

If you're eager to experiment with the Kotlin/Native SDK, you can try consuming artifacts from our test namespace: https://repo1.maven.org/maven2/com/sonatype/central/testing/amazon/. Note that these aren't officially supported and won't be updated frequently. If you do try them, we'd appreciate any feedback here or in a new issue!

**Comment by lauzadis** [comment]
I wanted to share another update here. We continue to experience performance issues with our publishing dependency, and we're still working with them towards resolution.

In the meantime, we're exploring alternatives to unblock this release, including self-hosting the artifacts in a custom AWS Maven repository. To consume K/N artifacts through this repository, you would need to [configure a custom repository](https://docs.gradle.org/current/userguide/declaring_repositories.html#sub:declaring-custom-repository-basics):
```kt
// build.gradle.kts
repositories {
    maven { // target AWS Maven repository for K/N artifacts
        url = uri("<TODO_AWS_MAVEN_REPO>")
    }
    mavenCentral() // fallback to Maven Central for all other dependencies
}
```

Please react to this comment with a 👍 if your project would accept this configuration. 

We understand some customers may be unable to use a custom Maven repository, and for you, we're continuing to work on getting these artifacts to Maven Central. Please react to this comment with a 👎 if your project would not support this type of configuration.

Thanks!

**Comment by steamstreetjon** [comment]
Forgive the pestering, but it's been a couple of months and just wanted to understand where we're at? The artifacts aren't being updated in the test repository anymore, so I'm concerned about relying on those. It seems like a private repo is a suitable solution to most and I'd love to have something semi official so I can begin work.

**Comment by steamstreetjon** [comment]
Three months later... any update? Getting close to assuming this is a dead project and moving to another platform.

**Comment by ianbotsf** [comment]
Sorry for the delay, we are still working towards supporting Kotlin/Native and hope to have something to show in the next month.
